MapboxCoreNavigation/OfflineService.swift

```diff
@@ -0,0 +1,379 @@
+
+     @param domain A flag indicating whether the Maps or Navigation pack should be canceled.
+     @param region The offline region to be canceled
+     */
+    public func cancelDownload(for region: OfflineRegion, domain: OfflineRegionDomain) {
+        instance?.cancelPackDownload(for: domain.commonDomain, metadata: region.region)
+    }
+
+    /**
+     @brief Starts the download of a pack
+
+     If a pack with the same metadata is already being downloaded, nothing will happen. If the metadata is different,
+     that download will be canceled, and a new download will be started. If the downloaded pack with this ID is
+     already complete, a new download will be started if the revision is different.
+
+     @param domain A flag indicating whether the Maps or Navigation pack should be downloaded.
+     If no value or nil is provided, packs for both domains will be downloaded
+     @param region The offline region to be downloaded
+     */
+    public func download(region: OfflineRegion, forDomain domain: OfflineRegionDomain? = nil) {
+        if let domainToDownload = domain {
+            instance?.downloadPack(for: domainToDownload.commonDomain, metadata: region.region)
+        } else {
+            if region.mapsPack != nil {
+                instance?.downloadPack(for: .maps, metadata: region.region)
+            }
+            if region.navigationPack != nil {
+                instance?.downloadPack(for: .navigation, metadata: region.region)
+            }
+        }
+    }
+
+    /**
+     @brief Add an observer that gets events as region state changes.
+
+     @param observer An object implementing the observer interface.
+     */
+    public func register(observer: OfflineServiceObserver) {
+        observers.append(observer)
+        notifyAvailable(forObserver: observer)
+    }
+
+    /**
+     @brief Removes an observer.
+
+     @param observer The observer that should be removed.
+     */
+    public func unregister(observer: OfflineServiceObserver) {
+        observers.removeAll {
+            $0 === observer
+        }
+    }
+
+    private func notifyAvailable(forObserver observer: OfflineServiceObserver) {
+        DispatchQueue.main.async { [weak self] in
+            self?.regions.forEach {
+                if $0.mapsPack != nil {
+                    observer.didBecomeAvailable(region: $0, forDomain: .maps)
+                }
+                if $0.navigationPack != nil {
+                    observer.didBecomeAvailable(region: $0, forDomain: .navigation)
+                }
+            }
+        }
+    }
+}
+
+extension OfflineService: MapboxCommon.OfflineServiceObserver {
+    public func onPending(for domain: OfflineDataDomain, metadata: OfflineDataRegionMetadata, pack: OfflineDataPack) {
+        DispatchQueue.main.async { [weak self] in
+            guard let self = self else { return }
+            let region = self._regions[metadata.id] ?? OfflineRegion(region: metadata)
```

Comment:
Is `metadata.id` really reliable way of using it as key in a dictionary? As I can see it contains names like `austin_day` which can be controlled by user during creation, this will in turn will lead to clash in regions which have the same id, but different content. I'd perform more testing with different ids to make sure.

Reply:
@kkaefer are the ids not unique for the regions?